### ISSUE TITLE
C++ Unit Testing Framework...

### ISSUE THREAD

**Issue by datahead8888** [Secretchronicles/TSC#427]
This has been discussed for a long time, and it's time to make a task for it.
This is a task to make a unit testing framework that tests different modules of TSC, each with a pass/fail status.
Tasks:
- Choose a unit testing API
- Decide what modules to test (For now)
  - Quintus has suggested file loading / saving
    - Game saves / loads
    - Level/World saving/loading
    - Campaign Loading
    - Potentially any config file usage
  - We might add unit tests for the C++ side of the level scripting Ruby API
  - We may think what else might benefit from testing.  There are probably additional interfaces that will benefit.
- We might talk more about how to deal with this in code reviews.  Many contributors will not write unit tests (or spend much time on them).  I know this from experience.

I'm a bit concerned with how we'll keep these up to date with many different contributors, but I do understand the benefits (how many times have save bugs been discovered that no one knew about for a long time?).  Feel free to share any thoughts on this task.

**Comment by datahead8888** [comment]
Last we discussed this, we seemed to be leaning toward not writing unit tests for TSC.  Should we close this, perhaps with something like "WONTFIX"?

**Comment by kirbyfan64** [comment]
Maybe at least some simple things could get tested?

**Comment by datahead8888** [comment]
@kirbyfan64, these were my arguments:
* It's unrealistic to expect that all people who come onto the project (or just file a pull request) will update the unit tests for each code change.  I've seen people slop through unit tests in corporate jobs, and it will be even worse in an open source project.
* If the unit tests are not updated, it will fall onto the core team members to get them up to date, which is frustrating
* In my view, having to maintain unit tests continually feels a little like corporate coding, since I've usually been asked to do work with them in corporate jobs
* It may take a reasonable investment of time before we start yielding notable benefits from the unit tests

Granted, I have seen some code recently that would have benefited quite a bit from more testing; I just don't think unit test scripts are the answer.  Feel free to offer other thoughts, though.

**Comment by kirbyfan64** [comment]
Well, I remember reading once how someone used assertions in code for "tests", just to check some invariants.

**Comment by Quintus** [comment]
I second all of datahead's points. Additionally, testing gameplay is
quite difficult. I have a hard time imagining how I would test whether
jumping with a certain speed allows crossing an abyss of a certain width
for example.

As for assertions: That's not really "unit" testing. It's some kind of
testing, but it serves a different purpose than unit testing. Latter are
meant to catch bugs before they even promote to a stable release, former
are there to improve bug reports made by users as they allow the
developer to find out what went wrong without having to fire up a
debugger, obtaining a backtrace, and seeing which invariant was
violated. Unfortunately, the information given by failing assertions
still does often not suffice, and you still have to fire up a debugger.

Also, assert() is for C code. In C++ we have exceptions, which can serve
(amongst other purposes) the same purpose, but are better because you
can try/catch them and handle the violation of an invariant. assert()
unconditionally terminates your programme.

Vale,
Quintus

-- 
Blog: http://www.guelkerdev.de
PGP/GPG ID: F1D8799FBCC8BC4F

**Comment by datahead8888** [comment]
> Also, assert() is for C code. In C++ we have exceptions, which can serve (amongst other purposes) the same purpose, but are better because you can try/catch them and handle the violation of an invariant.

The assertions are in the unit test code itself, not in the program being tested.  In the case of the Boost C++ testing library, you use methods such as BOOST_CHECK in order to run tests:
http://www.boost.org/doc/libs/1_42_0/libs/test/doc/html/tutorials/hello-the-testing-world.html

I've seen similar assertion mechanisms in C# and Java unit test frameworks.  You can, however, program a unit test to look for C++ exceptions, as with any other type of issue.

> As for assertions: That's not really "unit" testing.

It's all a question of nomenclature, but you always must do manual tests in the application in question; you can't assume it's tested because the unit test scripts pass.

> Unfortunately, the information given by failing assertions still does often not suffice, and you still have to fire up a debugger.

Yes, this is true, especially if you are not familiar with the unit test code or the code it is testing.  People also tend to be bad about including code comments in unit tests, I've found.

I have no doubt that unit tests can find problems in code, and they tend to force a different, rigorous methodology in programming, resulting in better thought out code.  Still, though, I stand by all the points I made above.  I think they will be difficult to support for TSC in the long term, and I don't think the investment versus benefit ratio warrants doing them.  Other thoughts on this are welcome, though.

**Comment by Quintus** [comment]
Since nobody started on this and I haven't changed my mind on that testing gameplay is difficult, I close this. I have also another point. We simply don't have the manpower to consequently write tests for everything.
